### Title: 
# Exploratory Data Analysis

### Description:
We are going to perform an EDA of our dataset.

### Authors:
#### Hugo Cesar Octavio del Sueldo
#### Jose Lopez Galdon

### Date:
23/10/2020

### Version:
1.0

***

### Libraries

We are going to start this EDA loading the respective libraries

In [3]:
    # Numpy & Pandas to work with the DF
import numpy as np
import pandas as pd

    # Seaborn / matplotlib for graphs
import seaborn as sns
import matplotlib.pyplot as plt 
%matplotlib inline

    # Automatic reports
from pandas_profiling import ProfileReport

    # Visualize DF
from IPython.display import display, HTML

***

## Load data

In [4]:
    # To automate the work as much as possible, we will parameterize the codes, so in this case, we will create an objetct with
    # the path root
name = 'data_train_merge'

data = pd.read_csv(f'../data/02_intermediate/{name}.csv',  # Path root: here we include an f-string with the variable name to parametrize
                   low_memory = False)                     # To avoid warnings we use set low_memory = False

***

## Variable predict

Taking into account the goals of the practice, we only want to know those persons that `Fully paid` their loans or those who does not, so we will tansform de dependent variable into binary problem, to ones or ceros depending on that condition. 

In [10]:
    # We will only select those observations with "Fully Paid" & "Charged Off"
data_binary = data[(data['loan_status'] == "Fully Paid") | (data['loan_status'] == "Charged Off")]

    # Now, we will transform into 0 & 1
dummy_dict = {"Fully Paid":0, "Charged Off":1}

    # Finally, we use the dictiony in the dataset
data = data_binary.replace({"loan_status": dummy_dict})

In [13]:
data

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
8       112038251        NaN      11575        11575          11575.0   
10      112149045        NaN       7200         7200           7200.0   
24      112052261        NaN       7500         7500           7500.0   
42      111999259        NaN      10000        10000          10000.0   
91      111808508        NaN      14000        14000          14000.0   
...           ...        ...        ...          ...              ...   
884869   36221446        NaN      11575        11575          11575.0   
884870   36330911        NaN      12000        12000          12000.0   
884873   36271333        NaN      13000        13000          13000.0   
884874   36490806        NaN      12000        12000          12000.0   
884875   36271262        NaN      20000        20000          20000.0   

              term int_rate  installment grade sub_grade  ...  \
8        36 months    7.35%       359.26     A        A4  ...   
10       36 months   24.85%       285.70     E        E3  ...   
24       36 months    7.35%       232.79     A        A4  ...   
42       60 months   16.02%       243.29     C        C5  ...   
91       36 months   16.02%       492.34     C        C5  ...   
...            ...      ...          ...   ...       ...  ...   
884869   36 months   15.59%       404.61     D        D1  ...   
884870   36 months   11.99%       398.52     B        B5  ...   
884873   60 months   15.99%       316.07     D        D2  ...   
884874   60 months   19.99%       317.86     E        E3  ...   
884875   36 months   11.99%       664.20     B        B5  ...   

       hardship_payoff_balance_amount hardship_last_payment_amount  \
8                                 NaN                          NaN   
10                                NaN                          NaN   
24                                NaN                          NaN   
42                                NaN                          NaN   
91                                NaN                          NaN   
...                               ...                          ...   
884869                            NaN                          NaN   
884870                            NaN                          NaN   
884873                            NaN                          NaN   
884874                            NaN                          NaN   
884875                            NaN                          NaN   

       disbursement_method  debt_settlement_flag debt_settlement_flag_date  \
8                     Cash                     N                       NaN   
10                    Cash                     N                       NaN   
24                    Cash                     N                       NaN   
42                    Cash                     N                       NaN   
91                    Cash                     N                       NaN   
...                    ...                   ...                       ...   
884869                Cash                     N                       NaN   
884870                Cash                     N                       NaN   
884873                Cash                     N                       NaN   
884874                Cash                     N                       NaN   
884875                Cash                     N                       NaN   

       settlement_status  settlement_date settlement_amount  \
8                    NaN              NaN               NaN   
10                   NaN              NaN               NaN   
24                   NaN              NaN               NaN   
42                   NaN              NaN               NaN   
91                   NaN              NaN               NaN   
...                  ...              ...               ...   
884869               NaN              NaN               NaN   
884870               NaN              NaN               NaN   
884873      

### View data

Firstly, we are goint to take a look to our dataframe:

In [14]:
    # First 10 rows using html display in order to view all the columns
data = data
display(HTML(data.head(10).to_html()))

In [15]:
    # Last 10 rows using html display in order to view all the columns
data = data
display(HTML(data.tail(10).to_html()))

In [5]:
    # Data dimension
data.shape

(884876, 151)

### Data wrangling

Once we have viewed our dataset, we continue exploring the column types and nulls...

In [6]:
data.info(verbose = True, null_counts = True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 884876 entries, 0 to 884875
Data columns (total 151 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   id                                          884876 non-null  int64  
 1   member_id                                   0 non-null       float64
 2   loan_amnt                                   884876 non-null  int64  
 3   funded_amnt                                 884876 non-null  int64  
 4   funded_amnt_inv                             884876 non-null  float64
 5   term                                        884876 non-null  object 
 6   int_rate                                    884876 non-null  object 
 7   installment                                 884876 non-null  float64
 8   grade                                       884876 non-null  object 
 9   sub_grade                                   884876 non-null  object 


As we can see above the only one column that has only nulls is `member_id`, therefore we will proceed to drop it. Also we observed that we have multiple columns with NaN, such as `desc`, `open_il_12m`, `open_rv_12m`...

After inspection we conclude that the following columns are not important for the analysis. We will describe why below.
- `url`: This variable does not provide relevant information because the url does not include information of the borrower.
- `desc`: As this variable is plenty of null and is very similar to `purpose` we will drop it.
- `title`: Has the same information as `purpose`.
- `open_acc_6m`, `open_act_il`...: According to some authors from *Towards Data Science* it is a good practice to drop those columns that has more than the 50% with NaN.

In [15]:
    # Drop columns
data = data.drop(columns = ['member_id', 'url', 'desc', 'title', 'open_acc_6m', 'open_act_il', 'open_il_12m', 'open_il_24m', 
                            'mths_since_rcnt_il', 'total_bal_il', 'il_util', 'open_rv_12m',  'open_rv_24m', 'max_bal_bc', 
                            'all_util', 'num_tl_120dpd_2m', 'annual_inc_joint', 'dti_joint', 'verification_status_joint', 
                            'last_pymnt_d', 'next_pymnt_d'])

In [16]:
    # Drop rows
data = data.dropna(subset = ['emp_title', 'emp_length', 'last_credit_pull_d', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util', 
                             'mo_sin_old_il_acct', 'mths_since_recent_bc', 'num_rev_accts', 'dti'])

In [17]:
    # Fill nulls with 0
        # First  of all we create an object call columns that include all the variables we will fill 0 value
columns = ['mths_since_last_delinq','mths_since_last_record', 'mths_since_last_major_derog', 'inq_fi', 'total_cu_tl', 
           'inq_last_12m', 'mths_since_recent_bc_dlq', 'mths_since_recent_inq', 'mths_since_recent_revol_delinq', 
           'percent_bc_gt_75', 'revol_bal_joint', 'sec_app_fico_range_low', 'sec_app_fico_range_high', 
           'sec_app_earliest_cr_line', 'sec_app_inq_last_6mths', 'sec_app_mort_acc', 'sec_app_open_acc', 
           'sec_app_revol_util', 'sec_app_open_act_il', 'sec_app_num_rev_accts', 'sec_app_chargeoff_within_12_mths', 
           'sec_app_collections_12_mths_ex_med', 'sec_app_mths_since_last_major_derog', 'hardship_type', 
           'hardship_reason', 'hardship_status', 'deferral_term',  'hardship_amount', 'hardship_start_date', 
           'hardship_end_date', 'payment_plan_start_date', 'hardship_length', 'hardship_dpd', 'hardship_loan_status', 
           'orig_projected_additional_accrued_interest', 'hardship_payoff_balance_amount',  
           'hardship_last_payment_amount', 'debt_settlement_flag_date', 'settlement_status',  'settlement_date',  
           'settlement_amount', 'settlement_percentage', 'settlement_term']
        
data[columns] = data[columns].fillna(value = 0)

In [18]:
data.shape

(801087, 130)

Now our dataset has 130 columns instead of 151.

Next, we will take a look at duplicates, so if we find something we will delete it.

In [8]:
data.duplicated().sum()

0

As we can see, we have any duplicated data.

In [25]:
data = data
display(HTML(data.tail(10).to_html()))

In [5]:
data['loan_status'].unique()

array(['Current', 'Fully Paid', 'In Grace Period', 'Late (31-120 days)',
       'Late (16-30 days)', 'Charged Off', 'Default'], dtype=object)